libzfs: add v2 iterator interfaces

robn opened the pull request:
### Motivation and Context

f6a0dac84 modified the zfs_iter_* functions to take a new "flags" parameter, and introduced a variety of flags to ask the kernel to limit the results in various ways, reducing the amount of work the caller needed to do to filter out things they didn't need.

Unfortunately this change broke the ABI for existing clients (read: older versions of the `zfs` program), and was reverted 399b98198.

dc95911d2 reintroduced the original patch, with the understanding that a backwards-compatible fix would be made before the 2.2 release branch was tagged. This commit is that fix.

### Description

This introduces zfs_iter_*_v2 functions that have the new flags argument, and reverts the existing functions to not have the flags parameter, as they were before. The old functions are now reimplemented in terms of the new, with flags set to 0.

This patch was originally grwilson/openzfs@197203719. I've cleaned up the formatting, updated the ABI reference, and given it some light testing.

### How Has This Been Tested?

I installed Debian `zfsutils-linux` 2.0.3 (as shipped with Debian 11) and ran various incantations of `zfs list` against a current kernel with this patch applied. Everything worked as I'd expect.

Automated tests should find something if there is anything.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

robn committed:
libzfs: add v2 iterator interfaces

f6a0dac84 modified the zfs_iter_* functions to take a new "flags"
parameter, and introduced a variety of flags to ask the kernel to limit
the results in various ways, reducing the amount of work the caller
needed to do to filter out things they didn't need.

Unfortunately this change broke the ABI for existing clients (read:
older versions of the `zfs` program), and was reverted 399b98198.

dc95911d2 reintroduced the original patch, with the understanding that a
backwards-compatible fix would be made before the 2.2 release branch was
tagged. This commit is that fix.

This introduces zfs_iter_*_v2 functions that have the new flags
argument, and reverts the existing functions to not have the flags
parameter, as they were before. The old functions are now reimplemented
in terms of the new, with flags set to 0.

Sponsored-by: Klara, Inc.
Original-patch-by: George Wilson <george.wilson@delphix.com>
Signed-off-by: Rob Norris <rob.norris@klarasystems.com>

ryao reviewed (commented):
I do not have time to review this PR this week (and I have a few PRs ahead of this in my review queue), but since this involves a library interface change, I suggest that there be at least a 2 week review period for this before a merge is done. That should give myself and others time to look at this in greater detail before we commit to supporting a library API extension forever.

grwilson commented:
I'm trying this out in our internal testing. It hit a few failures which I will investigate and provide feedback. Thanks for taking this on.

robn commented:
@grwilson any updates? I'd be happy to help look into something if that's useful?

grwilson commented on include/libzfs.h:
```diff
@@ -656,17 +656,29 @@ _LIBZFS_H void zprop_print_one_property(const char *, zprop_get_cbdata_t *,
-_LIBZFS_H int zfs_iter_filesystems(zfs_handle_t *, int, zfs_iter_f, void *);
-_LIBZFS_H int zfs_iter_snapshots(zfs_handle_t *, int, zfs_iter_f, void *,
+_LIBZFS_H int zfs_iter_filesystems(zfs_handle_t *, zfs_iter_f, void *);
+_LIBZFS_H int zfs_iter_snapshots(zfs_handle_t *, zfs_iter_f, void *,
```
This should be changed to:
```
_LIBZFS_H int zfs_iter_snapshots(zfs_handle_t *, boolean_t, zfs_iter_f, void *,
```
We need to add the `boolean_t` arg back.

grwilson commented on lib/libzfs/libzfs_iter.c:
```diff
@@ -103,7 +103,14 @@ zfs_do_list_ioctl(zfs_handle_t *zhp, int arg, zfs_cmd_t *zc)
  */
 int
-zfs_iter_filesystems(zfs_handle_t *zhp, int flags, zfs_iter_f func, void *data)
+zfs_iter_filesystems(zfs_handle_t *zhp, zfs_iter_f func, void *data)
```
Change signature to match original:
```
 int
 zfs_iter_snapshots(zfs_handle_t *zhp, boolean_t simple, zfs_iter_f func,
     void *data, uint64_t min_txg, uint64_t max_txg)
```

grwilson commented on lib/libzfs/libzfs_iter.c:
```diff
@@ -103,7 +103,14 @@ zfs_do_list_ioctl(zfs_handle_t *zhp, int arg, zfs_cmd_t *zc)
-zfs_iter_filesystems(zfs_handle_t *zhp, int flags, zfs_iter_f func, void *data)
+zfs_iter_filesystems(zfs_handle_t *zhp, zfs_iter_f func, void *data)
+{
+	return (zfs_iter_filesystems_v2(zhp, 0, func, data));
```
With the `boolean_t` arg added back then this could be changed to:
```
         return (zfs_iter_snapshots_v2(zhp, simple ? ZFS_ITER_SIMPLE : 0,
             func, data, min_txg, max_txg));
````

grwilson replied:
I'm running a test with these changes to verify that everything is working on our end.

robn replied:
Good spot, all taken as suggested.

robn commented:
Took the `zfs_iter_snapshots` fix suggested above, and rebased to latest. I've removed the `libzfs.abi` change for the moment as its difficult to rebase; I'll regenerate it and push it once we're settled.

grwilson commented:
I'll do another round of testing and then should be good to go.

robn commented:
> I'll do another round of testing and then should be good to go.

Thanks for that!

behlendorf reviewed (approved):
Looks like this just needs an updated ABI file.

robn committed:
libzfs: update ABI description for v2 iterators

Sponsored-by: Klara, Inc.
Signed-off-by: Rob Norris <rob.norris@klarasystems.com>

robn commented:
@behlendorf done.
